Added dump referenced objects command

En3Tho opened the pull request:
This is something I've found useful in my latest investigation. I've encountered a cache in the library we've used which happened ot be a "dominator" e.g. referenced many many objects. It was useful to know which objects were referenced. I think it can also be generally applied to get information on how much stuff a single object is transivitely referencing.

I wanted to make a DumpDominators command to automate such analysis (e.g. find objects that keep many other objects alive) but I'm still not sure how to do it. Calculations can take forever and there is a need to filter out objects that aren't really interesting (e.g. it has no sense to report static root as a dominator, it's obvious).

Also added an option to limit gcroots count as there were 403 roots of an object I tried to analyze. Generally I don't think there is a need to print that many roots so a limiting option is quite welcome.

Feel free to close this if you don't think this is helpful or have other future plans of similar work

leculver commented:
Doesn't `!objsize -stat` do the same thing?  Or am I misunderstanding?

https://github.com/dotnet/diagnostics/blob/main/src/Microsoft.Diagnostics.ExtensionCommands/ObjSizeCommand.cs

It currently doesn't accept a methodtable, but we could add a methodtable parameter to it to say "dump the objsize of all types of the given MT".

En3Tho commented:
@leculver Thank you for the review. Yes, objsize is doing the same thing and it was silly of me to miss that. I like the addition of method table parameter, I think it's useful. I will modify objsize and tag you when I'm done

leculver commented:
Yep, sounds good.

En3Tho commented:
@leculver PTAL. I added mt handling to objsize. Also, fixed few small things: moved space when writing out type so there is a space when type is not null and also added handling of `Short` parameter. It was unused.

leculver reviewed (approved):
LGTM, thanks for the code submission!
